File: lernfeld_04_datenbanken/python/loesungen/aufgabe_03.py
"""Aufgabe 3: Suchen & Sortieren mit SQL — Musterlösung (Python).

Such- und Sortierprogramm für die Notizen-Datenbank aus Aufgabe 1/2:
Alle Notizen (neueste zuerst), Stichwortsuche im Titel und LIMIT-Abfrage.
"""

import sqlite3


def alle_neueste(cur: sqlite3.Cursor) -> None:
    """Menüpunkt 1: Alle Notizen, neueste zuerst (erstellt_am absteigend)."""
    cur.execute(
        "SELECT id, titel, erstellt_am FROM notizen ORDER BY erstellt_am DESC"
    )
    zeilen = cur.fetchall()
    if not zeilen:
        print("Keine Notizen vorhanden.")
        return
    for nid, titel, datum in zeilen:
        print(f"[{nid}] {titel} – {datum}")


def stichwort_suche(cur: sqlite3.Cursor) -> None:
    """Menüpunkt 2: Alle Notizen, in deren Titel das Stichwort vorkommt."""
    wort = input("Stichwort: ").strip()
    # Die %-Wildcards stecken im Platzhalter-Wert, nicht im SQL-String
    muster = f"%{wort}%"

    # Trefferzahl zuerst melden
    cur.execute("SELECT COUNT(*) FROM notizen WHERE titel LIKE ?", (muster,))
    anzahl = cur.fetchone()[0]
    if anzahl == 0:
        print("Keine Treffer.")
        return
    print(f"{anzahl} Treffer:")

    cur.execute(
        "SELECT id, titel, erstellt_am FROM notizen "
        "WHERE titel LIKE ? ORDER BY erstellt_am DESC",
        (muster,),
    )
    for nid, titel, datum in cur.fetchall():
        print(f"[{nid}] {titel} – {datum}")


def neueste_n(cur: sqlite3.Cursor) -> None:
    """Menüpunkt 3: Nur die N neuesten Notizen (LIMIT)."""
    try:
        n = int(input("Anzahl: "))
    except ValueError:
        print("Ungültige Eingabe – bitte eine Zahl eingeben.")
        return
    if n < 1:
        print("Die Anzahl muss mindestens 1 sein.")
        return

    cur.execute(
        "SELECT id, titel, erstellt_am FROM notizen "
        "ORDER BY erstellt_am DESC LIMIT ?",
        (n,),
    )
    for nid, titel, datum in cur.fetchall():
        print(f"[{nid}] {titel} – {datum}")


def main() -> None:
    con = sqlite3.connect("notizen.db")
    cur = con.cursor()

    print("--- Notizen-Suche ---")
    while True:
        print("1: Alle Notizen (neueste zuerst)")
        print("2: Nach Stichwort suchen")
        print("3: Nur die neuesten N Notizen")
        print("0: Beenden")
        try:
            wahl = int(input("Wahl: "))
        except ValueError:
            print("Ungültige Eingabe – bitte eine Zahl eingeben.")
            continue

        if wahl == 1:
            alle_neueste(cur)
        elif wahl == 2:
            stichwort_suche(cur)
        elif wahl == 3:
            neueste_n(cur)
        elif wahl == 0:
            print("Auf Wiedersehen!")
            break
        else:
            print("Unbekannte Wahl – bitte 0–3 eingeben.")

    con.close()


if __name__ == "__main__":
    main()

File: lernfeld_04_datenbanken/cpp/loesungen/aufgabe_03.cpp
// Aufgabe 3: Suchen & Sortieren mit SQL — Musterlösung (C++)
//
// Such- und Sortierprogramm für die Notizen-Datenbank aus Aufgabe 1/2:
// Alle Notizen (neueste zuerst), Stichwortsuche im Titel und LIMIT-Abfrage.
//
// Kompilieren:  g++ -std=c++17 -Wall -Wextra aufgabe_03.cpp -o aufgabe_03 -lsqlite3
// Ausführen:    ./aufgabe_03

#include <iostream>  // std::cout, std::cin, std::getline
#include <limits>    // std::numeric_limits
#include <string>    // std::string

#include <sqlite3.h>

// Liest eine Ganzzahl ein und verwirft den Rest der Zeile.
bool int_lesen(int& wert) {
    std::cin >> wert;
    if (std::cin.fail()) {
        std::cin.clear();
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        wert = -1;   // C++ setzt den Wert sonst auf 0 – das wäre "Beenden"!
        return false;
    }
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    return true;
}

// Gibt eine Ergebniszeile (id, titel, erstellt_am) im gewohnten Format aus.
void zeile_ausgeben(sqlite3_stmt* stmt) {
    const int id = sqlite3_column_int(stmt, 0);
    const std::string titel(
        reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1)));
    const std::string datum(
        reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2)));
    std::cout << "[" << id << "] " << titel << " – " << datum << std::endl;
}

// Menüpunkt 1: Alle Notizen, neueste zuerst (erstellt_am absteigend).
void alle_neueste(sqlite3* db) {
    sqlite3_stmt* stmt = nullptr;
    sqlite3_prepare_v2(
        db, "SELECT id, titel, erstellt_am FROM notizen "
            "ORDER BY erstellt_am DESC;",
        -1, &stmt, nullptr);
    bool irgendwas = false;
    while (sqlite3_step(stmt) == SQLITE_ROW) {
        irgendwas = true;
        zeile_ausgeben(stmt);
    }
    if (!irgendwas) {
        std::cout << "Keine Notizen vorhanden." << std::endl;
    }
    sqlite3_finalize(stmt);
}

// Menüpunkt 2: Alle Notizen, in deren Titel das Stichwort vorkommt.
void stichwort_suche(sqlite3* db) {
    std::string wort;
    std::cout << "Stichwort: ";
    std::getline(std::cin, wort);
    // Die %-Wildcards stecken im gebundenen Wert, nicht im SQL-String!
    const std::string muster = "%" + wort + "%";

    // Trefferzahl zuerst melden
    sqlite3_stmt* stmt = nullptr;
    sqlite3_prepare_v2(
        db, "SELECT COUNT(*) FROM notizen WHERE titel LIKE ?;",
        -1, &stmt, nullptr);
    sqlite3_bind_text(stmt, 1, muster.c_str(), -1, SQLITE_TRANSIENT);
    int anzahl = 0;
    if (sqlite3_step(stmt) == SQLITE_ROW) {
        anzahl = sqlite3_column_int(stmt, 0);
    }
    sqlite3_finalize(stmt);

    if (anzahl == 0) {
        std::cout << "Keine Treffer." << std::endl;
        return;
    }
    std::cout << anzahl << " Treffer:" << std::endl;

    sqlite3_prepare_v2(
        db, "SELECT id, titel, erstellt_am FROM notizen "
            "WHERE titel LIKE ? ORDER BY erstellt_am DESC;",
        -1, &stmt, nullptr);
    sqlite3_bind_text(stmt, 1, muster.c_str(), -1, SQLITE_TRANSIENT);
    while (sqlite3_step(stmt) == SQLITE_ROW) {
        zeile_ausgeben(stmt);
    }
    sqlite3_finalize(stmt);
}

// Menüpunkt 3: Nur die N neuesten Notizen (LIMIT mit Platzhalter).
void neueste_n(sqlite3* db) {
    int n;
    if (!int_lesen(n)) {
        std::cout << "Ungültige Eingabe – bitte eine Zahl eingeben."
                  << std::endl;
        return;
    }
    if (n < 1) {
        std::cout << "Die Anzahl muss mindestens 1 sein." << std::endl;
        return;
    }

    sqlite3_stmt* stmt = nullptr;
    sqlite3_prepare_v2(
        db, "SELECT id, titel, erstellt_am FROM notizen "
            "ORDER BY erstellt_am DESC LIMIT ?;",
        -1, &stmt, nullptr);
    sqlite3_bind_int(stmt, 1, n);
    while (sqlite3_step(stmt) == SQLITE_ROW) {
        zeile_ausgeben(stmt);
    }
    sqlite3_finalize(stmt);
}

int main() {
    sqlite3* db = nullptr;
    int rc = sqlite3_open("notizen.db", &db);
    if (rc != SQLITE_OK) {
        std::cerr << "Fehler beim Öffnen der Datenbank: "
                  << sqlite3_errmsg(db) << std::endl;
        sqlite3_close(db);
        return 1;
    }

    std::cout << "--- Notizen-Suche ---" << std::endl;
    int wahl = -1;
    while (wahl != 0) {
        std::cout << "1: Alle Notizen (neueste zuerst)" << std::endl;
        std::cout << "2: Nach Stichwort suchen" << std::endl;
        std::cout << "3: Nur die neuesten N Notizen" << std::endl;
        std::cout << "0: Beenden" << std::endl;
        std::cout << "Wahl: ";
        if (!int_lesen(wahl)) {
            std::cout << "Ungültige Eingabe – bitte eine Zahl eingeben."
                      << std::endl;
            continue;
        }
        switch (wahl) {
            case 1:  alle_neueste(db); break;
            case 2:  stichwort_suche(db); break;
            case 3:  neueste_n(db); break;
            case 0:
                std::cout << "Auf Wiedersehen!" << std::endl;
                break;
            default:
                std::cout << "Unbekannte Wahl – bitte 0–3 eingeben."
                          << std::endl;
        }
    }

    sqlite3_close(db);
    return 0;
}
